pilot/pkg/config/kube/gateway/conversion.go

```diff
@@ -1404,17 +1411,34 @@ func buildGRPCDestination(
 	return res, invalidBackendErr, nil
 }
 
-func buildDestination(ctx configContext, to k8s.BackendRef, ns string, enforceRefGrant bool) (*istio.Destination, *ConfigError) {
-	// check if the reference is allowed
-	if enforceRefGrant {
+type refGrantChecker func(ctx configContext, to k8s.BackendRef, ns string) error
+
+func newRefGrantChecker(enforceRefGrant bool, ks ...config.GroupVersionKind) refGrantChecker {
+	if !enforceRefGrant {
+		return func(_ configContext, _ k8s.BackendRef, _ string) error { return nil }
+	}
+	return func(ctx configContext, to k8s.BackendRef, ns string) error {
 		refs := ctx.AllowedReferences
 		if toNs := to.Namespace; toNs != nil && string(*toNs) != ns {
-			if !refs.BackendAllowed(gvk.HTTPRoute, to.Name, *toNs, ns) {
-				return &istio.Destination{}, &ConfigError{
-					Reason:  InvalidDestinationPermit,
-					Message: fmt.Sprintf("backendRef %v/%v not accessible to a route in namespace %q (missing a ReferenceGrant?)", to.Name, *toNs, ns),
+			var deniedKinds []string
```

Comment:
why do we need multiple? seems like we only ever use one, and I cannot imagine having multiple in the future